<!DOCTYPE html>
<html lang="en"><head>
<script src="atacseq_slides_files/libs/clipboard/clipboard.min.js"></script>
<script src="atacseq_slides_files/libs/quarto-html/tabby.min.js"></script>
<script src="atacseq_slides_files/libs/quarto-html/popper.min.js"></script>
<script src="atacseq_slides_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="atacseq_slides_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="atacseq_slides_files/libs/quarto-html/light-border.css" rel="stylesheet">
<link href="atacseq_slides_files/libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.7.30">

  <meta name="author" content="Sean Davis, MD, PhD">
  <meta name="dcterms.date" content="2025-06-17">
  <title>Chromatin Accessibility and ATAC-seq</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="atacseq_slides_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="atacseq_slides_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    html { -webkit-text-size-adjust: 100%; }
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
    /* CSS for citations */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
      margin-bottom: 0em;
    }
    .hanging-indent div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }  </style>
  <link rel="stylesheet" href="atacseq_slides_files/libs/revealjs/dist/theme/quarto-f1bc407cc908ae3bd40ff94d579acb78.css">
  <link href="atacseq_slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="atacseq_slides_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="atacseq_slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="atacseq_slides_files/libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="atacseq_slides_files/libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="atacseq_slides_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <script src="atacseq_slides_files/libs/quarto-diagram/mermaid.min.js"></script>
  <script src="atacseq_slides_files/libs/quarto-diagram/mermaid-init.js"></script>
  <link href="atacseq_slides_files/libs/quarto-diagram/mermaid.css" rel="stylesheet">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" data-background-image="images/atacseq_title_bg.png" data-background-opacity="1" data-background-size="cover" class="quarto-title-block center">
  <h1 class="title">Chromatin Accessibility and ATAC-seq</h1>
  <p class="subtitle">A Primer for C-DATA-2025</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Sean Davis, MD, PhD 
</div>
        <p class="quarto-title-affiliation">
            University of Colorado Anschutz Medical Campus
          </p>
    </div>
</div>

  <p class="date">June 17, 2025</p>
</section>
<section>
<section id="measuring-chromatin-accessibility" class="title-slide slide level1 center">
<h1>Measuring Chromatin Accessibility</h1>

</section>
<section id="early-chromatin-accessibility" class="slide level2">
<h2>Early Chromatin Accessibility</h2>

<img data-src="boyle_dnase_title_page.png" class="r-stretch quarto-figure-center" id="fig-boyle-dnase-seq"><p class="caption">
Figure&nbsp;1: The first DNase-seq experiment identified open chromatin regions in a genome-wide manner. <span class="citation" data-cites="boyleHighresolutionMappingCharacterization2008">(<a href="#/references" role="doc-biblioref" onclick="">Boyle et al. 2008</a>)</span>
</p></section>
<section id="dnase-seq-signal" class="slide level2">
<h2>DNAse-seq Signal</h2>

<img data-src="2025-06-17-00-17-36.png" class="r-stretch quarto-figure-center" id="fig-dnase-seq-signal"><p class="caption">
Figure&nbsp;2: Successive zoom-in of original DNAse-seq experiment.
</p></section>
<section id="dnase-seq-peaks" class="slide level2">
<h2>DNAse-seq Peaks</h2>

<img data-src="2025-06-17-00-23-20.png" class="r-stretch quarto-figure-center" id="fig-dnase-seq-peaks"><p class="caption">
Figure&nbsp;3: Positions of DNase-seq peaks across the genome.
</p></section>
<section id="dnase-seq-peaks-and-signal" class="slide level2">
<h2>DNAse-seq Peaks and Signal</h2>

<img data-src="2025-06-17-00-25-12.png" class="r-stretch quarto-figure-center" id="fig-dnase-seq-peaks-and-expression"><p class="caption">
Figure&nbsp;4: Relationship of DNAse-seq peaks and gene expression.
</p></section>
<section id="chromatin-accessibility-profiling" class="slide level2">
<h2>Chromatin Accessibility Profiling</h2>
<table class="caption-top">
<colgroup>
<col style="width: 14%">
<col style="width: 23%">
<col style="width: 10%">
<col style="width: 30%">
<col style="width: 20%">
</colgroup>
<thead>
<tr class="header">
<th>Method</th>
<th>Cell Number</th>
<th>Time</th>
<th>Nucleosome Info</th>
<th>Reference</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>ATAC-seq</strong></td>
<td>50,000</td>
<td>~3 hours</td>
<td>✓</td>
<td><span class="citation" data-cites="buenrostroTranspositionNativeChromatin2013">(<a href="#/references" role="doc-biblioref" onclick="">Buenrostro et al. 2013</a>)</span></td>
</tr>
<tr class="even">
<td>DNase-seq</td>
<td>50 million</td>
<td>~3 days</td>
<td>✗</td>
<td><span class="citation" data-cites="boyleHighresolutionMappingCharacterization2008">(<a href="#/references" role="doc-biblioref" onclick="">Boyle et al. 2008</a>)</span></td>
</tr>
<tr class="odd">
<td>FAIRE-seq</td>
<td>25 million</td>
<td>~5 days</td>
<td>✗</td>
<td><span class="citation" data-cites="giresiIsolationActiveRegulatory2009">(<a href="#/references" role="doc-biblioref" onclick="">Giresi and Lieb 2009</a>)</span></td>
</tr>
</tbody>
</table>
</section></section>
<section>
<section id="introduction-to-atac-seq" class="title-slide slide level1 center">
<h1>Introduction to ATAC-Seq</h1>

</section>
<section id="section" class="slide level2">
<h2></h2>

<img data-src="buenrostro_cover_page.png" class="r-stretch quarto-figure-center"><p class="caption">ATAC-seq first described in <span class="citation" data-cites="buenrostroTranspositionNativeChromatin2013">Buenrostro et al. (<a href="#/references" role="doc-biblioref" onclick="">2013</a>)</span>.</p></section>
<section id="what-is-atac-seq" class="slide level2">
<h2>What is ATAC-seq?</h2>
<ul>
<li class="fragment"><strong>A</strong>ssay for <strong>T</strong>ransposase-<strong>A</strong>ccessible <strong>C</strong>hromatin using <strong>seq</strong>uencing</li>
<li class="fragment">Maps genome-wide chromatin accessibility</li>
<li class="fragment">Identifies open chromatin regions where regulatory proteins can bind</li>
<li class="fragment">Developed by Buenrostro et al.&nbsp;(2013) <span class="citation" data-cites="buenrostroTranspositionNativeChromatin2013">(<a href="#/references" role="doc-biblioref" onclick="">Buenrostro et al. 2013</a>)</span></li>
</ul>
</section>
<section id="atac-seq-overview" class="slide level2">
<h2>ATAC-Seq Overview</h2>

<img data-src="https://www.genewiz.com/wp-content/uploads/2024/07/ATAC-Seq_Overview.png" class="r-stretch quarto-figure-center" id="fig-atac-seq-sequencing"><p class="caption">
Figure&nbsp;5: Assay for transposase-accessible chromatin sequencing (ATAC-Seq) employs a hyperactive form of Tn5 transposase to identify regions of open chromatin, which are important for global epigenetic control of gene expression. Tn5 simultaneously cleaves and adds adapters to nucleosome-free regions of DNA, priming them for sequencing.
</p></section>
<section id="key-applications" class="slide level2">
<h2>Key Applications</h2>
<ul>
<li class="fragment"><strong>Regulatory element identification</strong>: Promoters, enhancers, silencers</li>
<li class="fragment"><strong>Transcription factor binding</strong>: Footprinting analysis</li>
<li class="fragment"><strong>Chromatin state mapping</strong>: Active vs inactive regions</li>
<li class="fragment"><strong>Developmental studies</strong>: Differentiation and cell fate decisions</li>
<li class="fragment"><strong>Disease research</strong>: Cancer epigenomics, genetic variants</li>
</ul>
</section></section>
<section>
<section id="data-analysis-pipeline" class="title-slide slide level1 center">
<h1>Data Analysis Pipeline</h1>

</section>
<section id="raw-data-processing" class="slide level2">
<h2>Raw Data Processing</h2>
<ol type="1">
<li class="fragment"><strong>Quality Assessment</strong>: FastQC, MultiQC</li>
<li class="fragment"><strong>Adapter Trimming</strong>: Trimmomatic, cutadapt</li>
<li class="fragment"><strong>Alignment</strong>: BWA-MEM, Bowtie2 <span class="citation" data-cites="langmead2012fast">(<a href="#/references" role="doc-biblioref" onclick=""><strong>langmead2012fast?</strong></a>)</span></li>
<li class="fragment"><strong>Post-alignment filtering</strong>: Remove duplicates, mitochondrial reads</li>
<li class="fragment"><strong>File format conversion</strong>: SAM → BAM → BED</li>
</ol>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href=""></a><span class="co"># Example alignment command</span></span>
<span id="cb1-2"><a href=""></a><span class="ex">bwa</span> mem <span class="at">-t</span> 8 genome.fa sample_R1.fastq.gz sample_R2.fastq.gz <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb1-3"><a href=""></a><span class="ex">samtools</span> sort <span class="at">-@</span> 8 <span class="at">-o</span> sample.bam</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section></section>
<section>
<section id="quality-control-metrics" class="title-slide slide level1 center">
<h1>Quality Control Metrics</h1>

</section>
<section id="quality-control-metrics-1" class="slide level2 smaller">
<h2>Quality Control Metrics</h2>
<ul>
<li>Experiments should have two or more biological replicates. Assays performed using EN-TEx samples may be exempted due to limited availability of experimental material, but at least two technical replicates are required.</li>
<li>Each replicate should have 25 million non-duplicate, non-mitochondrial aligned reads for single-end sequencing and 50 million for paired-ended sequencing (i.e.&nbsp;25 million fragments, regardless of sequencing run type).</li>
<li>The alignment rate, or percentage of mapped reads, should be greater than 95%, though values &gt;80% may be acceptable.</li>
<li>The percentage of reads mapping to the mitochondrial genome should be less than 10%.</li>
<li><a href="https://www.encodeproject.org/data-standards/terms/#enrichment">Transcription Start Site (TSS) enrichment</a></li>
</ul>
</section>
<section id="library-complexity" class="slide level2 smaller">
<h2>Library Complexity</h2>
<h3 id="pcr-bottlenecking-coefficient-1-pbc1">PCR Bottlenecking Coefficient 1 (PBC1)</h3>
<p>PBC1=<span class="math inline">\({M_1/M_{DISTINCT}}\)</span></p>
<p><span class="math inline">\(M_1\)</span>: number of genomic locations where exactly one read maps uniquely</p>
<p><span class="math inline">\(M_{DISTINCT}\)</span>: number of distinct genomic locations to which some read maps uniquely</p>
<h3 id="pcr-bottlenecking-coefficient-2-pbc2">PCR Bottlenecking Coefficient 2 (PBC2)</h3>
<p>PBC2=<span class="math inline">\(M_1/M_2\)</span></p>
<p><span class="math inline">\(M_1\)</span>: number of genomic locations where only one read maps uniquely</p>
<p><span class="math inline">\(M_2\)</span>: number of genomic locations where two reads map uniquely</p>
<h3 id="non-redundant-fraction-nrf">Non-Redundant Fraction (NRF)</h3>
<p>NRF = <span class="math inline">\({R_{distinct}}/{R_{total}}\)</span></p>
<p><span class="math inline">\(R_{distinct}\)</span>: number of distinct genomic locations to which some read maps uniquely</p>
<p><span class="math inline">\(R_{total}\)</span>: total number of reads</p>
</section>
<section id="common-pitfalls-and-solutions" class="slide level2">
<h2>Common Pitfalls and Solutions</h2>
<table class="caption-top">
<colgroup>
<col style="width: 34%">
<col style="width: 26%">
<col style="width: 38%">
</colgroup>
<thead>
<tr class="header">
<th>Problem</th>
<th>Cause</th>
<th>Solution</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Low library complexity</td>
<td>Over-amplification</td>
<td>Optimize PCR cycles</td>
</tr>
<tr class="even">
<td>High mitochondrial reads</td>
<td>Poor nuclei isolation</td>
<td>Improve protocol</td>
</tr>
<tr class="odd">
<td>Poor TSS enrichment</td>
<td>Degraded chromatin</td>
<td>Fresh samples, cold buffers</td>
</tr>
<tr class="even">
<td>Batch effects</td>
<td>Processing variation</td>
<td>Randomization, ComBat-seq</td>
</tr>
<tr class="odd">
<td>Low reproducibility</td>
<td>Technical noise</td>
<td>Increase replicates</td>
</tr>
</tbody>
</table>
</section></section>
<section>
<section id="core-analysis-approaches" class="title-slide slide level1 center">
<h1>Core Analysis Approaches</h1>

</section>
<section id="macs2-parameters-for-atac-seq" class="slide level2">
<h2>MACS2 Parameters for ATAC-seq</h2>
<p>MACS2 is a widely used peak calling tool for ATAC-seq data. The following parameters are commonly used:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href=""></a><span class="co"># Standard ATAC-seq peak calling</span></span>
<span id="cb2-2"><a href=""></a><span class="ex">macs2</span> callpeak <span class="at">-t</span> sample.bam <span class="dt">\</span></span>
<span id="cb2-3"><a href=""></a>               <span class="at">-f</span> BAMPE <span class="dt">\</span></span>
<span id="cb2-4"><a href=""></a>               <span class="at">-n</span> sample <span class="dt">\</span></span>
<span id="cb2-5"><a href=""></a>               <span class="at">-q</span> 0.01 <span class="dt">\</span></span>
<span id="cb2-6"><a href=""></a>               ...</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="read-shifting" class="slide level2">
<h2>Read shifting</h2>

<img data-src="https://media.springernature.com/m685/springer-static/image/art%3A10.1038%2Fs41596-022-00692-9/MediaObjects/41596_2022_692_Fig1_HTML.png" class="r-stretch quarto-figure-center"><p class="caption">The positive strand reads are shifted by 4 bp and the negative strand by 5 bp due to the 9 base offset introduced by Tn5 transposase.</p></section>
<section id="differential-accessibility-analysis" class="slide level2">
<h2>Differential Accessibility Analysis</h2>

<img data-src="https://stuartlab.org/signac/articles/pbmc_vignette_files/figure-html/unnamed-chunk-44-1.png" class="r-stretch quarto-figure-center" id="fig-differential-accessibility"><p class="caption">
Figure&nbsp;6: Differential Accessibility.
</p><p><strong>Differential binding detection</strong>: DESeq2, edgeR, csaw, DiffBind</p>
<div class="center">
<p>Replicates required!</p>
</div>
</section>
<section id="data-preparation" class="slide level2">
<h2>Data preparation</h2>
<ul>
<li class="fragment">Region creation</li>
<li class="fragment">Count matrix generation</li>
<li class="fragment">Normalization</li>
<li class="fragment">Differential analysis</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>          Sample_1 Sample_2 Sample_3 Sample_4 Sample_5
Region_1       210      974      375      404      109
Region_2       328      659      392      824      918
Region_3       617       48      638      904       68
Region_4       533       82      238      217      953
Region_5       883      666      960      642      597
Region_6       769      286      821      283      863
Region_7       265      509      172      357      625
Region_8       431      216      617       75      716
Region_9       329       69      142      713      970
Region_10      820      259      467       23       84</code></pre>
</div>
</div>
</section>
<section id="peak-annotation" class="slide level2">
<h2>Peak annotation</h2>

<img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1186%2F1471-2105-11-237/MediaObjects/12859_2009_Article_3694_Fig8_HTML.jpg" class="quarto-figure quarto-figure-center r-stretch" width="600"><ul>
<li><a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-237" class="uri">https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-237</a></li>
<li><a href="https://bioconductor.org/packages/ChipPeakAnno/">ChipPeakAnno</a> package</li>
</ul>
</section>
<section id="other" class="slide level2">
<h2>Other</h2>
<ul>
<li>Motif Analysis</li>
<li>Functional Annotation</li>
</ul>
<p><strong>Annotation Strategies</strong></p>
<ul>
<li><strong>Promoter assignment</strong>: TSS ± 2kb</li>
<li><strong>Enhancer prediction</strong>: Distal peaks + H3K27ac</li>
<li><strong>Gene assignment</strong>: Nearest gene, regulatory domains</li>
<li><strong>Pathway analysis</strong>: GO, KEGG, Reactome</li>
</ul>
</section></section>
<section>
<section id="single-cell-atac-seq-scatac-seq" class="title-slide slide level1 center">
<h1>Single-cell ATAC-seq (scATAC-seq)</h1>

</section>
<section id="overview-of-scatac-seq" class="slide level2">
<h2>Overview of scATAC-seq</h2>
<div class="cell" data-reveal="true" data-layout-align="center">
<div class="cell-output-display">
<div>
<p></p><figure class=""><p></p>
<div>
<pre class="mermaid mermaid-js">flowchart TD
    A[Single-cell ATAC-seq] --&gt; B[Cell Isolation]
    B --&gt; C[Transposition with Tn5]
    C --&gt; D[Library Preparation]
    D --&gt; E[Sequencing]
</pre>
</div>
<p></p><figcaption> Wet-lab workflow for single-cell ATAC-seq (scATAC-seq).</figcaption> </figure><p></p>
</div>
</div>
</div>
</section>
<section id="scatac-seq-workflow" class="slide level2">
<h2>scATAC-seq Workflow</h2>
<div class="cell" data-reveal="true" data-layout-align="center">
<div class="cell-output-display">
<div>
<p></p><figure class=""><p></p>
<div>
<pre class="mermaid mermaid-js">flowchart TD
    A[scATAC-seq Data] --&gt; B[Quality Control]
    B --&gt; C[Alignment]
    C --&gt; D[Binning and Counting]
    D --&gt; E[Cell Barcode Assignment]
    E --&gt; F[Dimensionality Reduction]
    F --&gt; G[Clustering]
</pre>
</div>
<p></p><figcaption> Computational workflow for single-cell ATAC-seq (scATAC-seq).</figcaption> </figure><p></p>
</div>
</div>
</div>
</section>
<section id="clustering" class="slide level2">
<h2>Clustering</h2>

<img data-src="https://www.activemotif.com/uploads/images/web_site/services-hp-dec2015/scATAC-kidney-figure-1-v2.png" class="r-stretch quarto-figure-center"><p class="caption">Assigning cell types to clusters involves “mapping” back into RNA space. Multiple approaches exist that utilize biological knowledge or ATAC-seq signal information.</p></section>
<section id="aggregating-reads-from-single-cells" class="slide level2">
<h2>Aggregating Reads from Single Cells</h2>

<img data-src="https://media.springernature.com/full/springer-static/image/art%3A10.1038%2Fs43586-020-00008-9/MediaObjects/43586_2020_8_Fig1_HTML.png" class="r-stretch quarto-figure-center"><p class="caption">Aggregating reads to generate “pseudobulk” samples used for peak-calling, differential analysis, and cell type identification.</p></section>
<section id="full-stack-analysis" class="slide level2">
<h2>Full stack analysis</h2>
<p><strong>Key Tools</strong>:</p>
<ul>
<li>ArchR</li>
<li>Signac</li>
<li>SnapATAC</li>
<li>And the list goes on…</li>
</ul>
</section></section>
<section>
<section id="practical-considerations" class="title-slide slide level1 center">
<h1>Practical Considerations</h1>

</section>
<section id="experimental-design-principles" class="slide level2">
<h2>Experimental Design Principles</h2>
<ul>
<li class="fragment"><strong>Biological replicates</strong>: Minimum 3 per condition</li>
<li class="fragment"><strong>Technical replicates</strong>: Generally not necessary</li>
<li class="fragment"><strong>Batch effects</strong>: Randomize sample processing</li>
<li class="fragment"><strong>Cell number</strong>: 50,000 optimal for bulk ATAC-seq</li>
</ul>
</section>
<section id="computational-infrastructure" class="slide level2">
<h2>Computational Infrastructure</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Hardware Requirements</strong></p>
<ul>
<li><strong>CPU</strong>: 16+ cores recommended</li>
<li><strong>RAM</strong>: 32+ GB for human genome</li>
<li><strong>Storage</strong>: 1+ TB for large projects</li>
<li><strong>GPU</strong>: Optional for deep learning</li>
</ul>
</div><div class="column" style="width:50%;">
<p><strong>Software Ecosystem</strong></p>
<ul>
<li><strong>Alignment</strong>: BWA-MEM, Bowtie2</li>
<li><strong>Peak calling</strong>: MACS2, HOMER</li>
<li><strong>Visualization</strong>: IGV, deepTools</li>
<li><strong>Analysis</strong>: R/Bioconductor, Python</li>
</ul>
</div></div>
</section>
<section id="workflow-management" class="slide level2">
<h2>Workflow Management</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href=""></a><span class="co"># Example Snakemake workflow</span></span>
<span id="cb4-2"><a href=""></a><span class="ex">rule</span> all:</span>
<span id="cb4-3"><a href=""></a>    <span class="ex">input:</span></span>
<span id="cb4-4"><a href=""></a>        <span class="st">"results/multiqc_report.html"</span><span class="ex">,</span></span>
<span id="cb4-5"><a href=""></a>        <span class="st">"results/differential_peaks.bed"</span></span>
<span id="cb4-6"><a href=""></a></span>
<span id="cb4-7"><a href=""></a><span class="ex">rule</span> trim_adapters:</span>
<span id="cb4-8"><a href=""></a>    <span class="ex">input:</span></span>
<span id="cb4-9"><a href=""></a>        <span class="va">r1</span><span class="op">=</span><span class="st">"data/{sample}_R1.fastq.gz"</span>,</span>
<span id="cb4-10"><a href=""></a>        <span class="va">r2</span><span class="op">=</span><span class="st">"data/{sample}_R2.fastq.gz"</span></span>
<span id="cb4-11"><a href=""></a>    <span class="ex">output:</span></span>
<span id="cb4-12"><a href=""></a>        <span class="va">r1</span><span class="op">=</span><span class="st">"trimmed/{sample}_R1_trimmed.fastq.gz"</span>,</span>
<span id="cb4-13"><a href=""></a>        <span class="va">r2</span><span class="op">=</span><span class="st">"trimmed/{sample}_R2_trimmed.fastq.gz"</span></span>
<span id="cb4-14"><a href=""></a>    <span class="ex">shell:</span></span>
<span id="cb4-15"><a href=""></a>        <span class="st">"trimmomatic PE {input.r1} {input.r2} {output.r1} {output.r2}"</span></span>
<span id="cb4-16"><a href=""></a></span>
<span id="cb4-17"><a href=""></a><span class="ex">rule</span> align:</span>
<span id="cb4-18"><a href=""></a>    <span class="ex">input:</span></span>
<span id="cb4-19"><a href=""></a>        <span class="va">r1</span><span class="op">=</span><span class="st">"trimmed/{sample}_R1_trimmed.fastq.gz"</span>,</span>
<span id="cb4-20"><a href=""></a>        <span class="va">r2</span><span class="op">=</span><span class="st">"trimmed/{sample}_R2_trimmed.fastq.gz"</span></span>
<span id="cb4-21"><a href=""></a>    <span class="ex">output:</span></span>
<span id="cb4-22"><a href=""></a>        <span class="st">"aligned/{sample}.bam"</span></span>
<span id="cb4-23"><a href=""></a>    <span class="ex">shell:</span></span>
<span id="cb4-24"><a href=""></a>        <span class="st">"bwa mem genome.fa {input.r1} {input.r2} | samtools sort -o {output}"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Tools</strong>: Snakemake, Nextflow, WDL <span class="citation" data-cites="Snakemak56:online">(<a href="#/references" role="doc-biblioref" onclick=""><span>“Snakemake–a Scalable Bioinformatics Workflow Engine - PubMed”</span> 2011</a>)</span></p>
</section>
<section id="reproducibility-best-practices" class="slide level2">
<h2>Reproducibility Best Practices</h2>
<ul>
<li class="fragment"><strong>Version control</strong>: Git for code, DVC for data</li>
<li class="fragment"><strong>Environment management</strong>: Conda, Docker, Singularity</li>
<li class="fragment"><strong>Documentation</strong>: Literate programming with R Markdown/Quarto</li>
<li class="fragment"><strong>Code sharing</strong>: GitHub, Zenodo for long-term archival</li>
<li class="fragment"><strong>Data repositories</strong>: GEO, SRA, ENCODE portal</li>
</ul>
</section>
<section id="online-resources" class="slide level2">
<h2>Online Resources</h2>
<ul>
<li><strong>ENCODE Portal</strong>: https://www.encodeproject.org/</li>
<li><strong>Bioconductor workflows</strong>: https://bioconductor.org/packages/ATACseqQC/</li>
<li><strong>Galaxy Training</strong>: https://training.galaxyproject.org/training-material/topics/epigenetics/</li>
<li><strong>nf-core/atacseq</strong>: https://nf-co.re/atacseq</li>
</ul>
</section>
<section id="references" class="slide level2 smaller scrollable">
<h2>References</h2>

<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-boyleHighresolutionMappingCharacterization2008" class="csl-entry" role="listitem">
Boyle, Alan P, Sean Davis, Hennady P Shulha, Paul Meltzer, Elliott H Margulies, Zhiping Weng, Terrence S Furey, and Gregory E Crawford. 2008. <span>“High-Resolution Mapping and Characterization of Open Chromatin Across the Genome.”</span> <em>Cell</em> 132 (2): 311–22. <a href="https://doi.org/10.1016/j.cell.2007.12.014">https://doi.org/10.1016/j.cell.2007.12.014</a>.
</div>
<div id="ref-buenrostroTranspositionNativeChromatin2013" class="csl-entry" role="listitem">
Buenrostro, Jason D, Paul G Giresi, Lisa C Zaba, Howard Y Chang, and William J Greenleaf. 2013. <span>“Transposition of Native Chromatin for Fast and Sensitive Epigenomic Profiling of Open Chromatin, <span class="nocase">DNA-binding</span> Proteins and Nucleosome Position.”</span> <em>Nature Methods</em> 10 (12): 1213–18. <a href="https://doi.org/10.1038/nmeth.2688">https://doi.org/10.1038/nmeth.2688</a>.
</div>
<div id="ref-giresiIsolationActiveRegulatory2009" class="csl-entry" role="listitem">
Giresi, Paul G., and Jason D. Lieb. 2009. <span>“Isolation of Active Regulatory Elements from Eukaryotic Chromatin Using <span>FAIRE</span> (<span>Formaldehyde Assisted Isolation</span> of <span>Regulatory Elements</span>).”</span> <em>Methods (San Diego, Calif.)</em> 48 (3): 233–39. <a href="https://doi.org/10.1016/j.ymeth.2009.03.003">https://doi.org/10.1016/j.ymeth.2009.03.003</a>.
</div>
<div id="ref-Snakemak56:online" class="csl-entry" role="listitem">
<span>“Snakemake–a Scalable Bioinformatics Workflow Engine - PubMed.”</span> 2011. <a href="https://pubmed.ncbi.nlm.nih.gov/22908215/">https://pubmed.ncbi.nlm.nih.gov/22908215/</a>.
</div>
</div>
</section></section>
    </div>
  <div class="quarto-auto-generated-content" style="display: none;">
<div class="footer footer-default">

</div>
</div></div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="atacseq_slides_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="atacseq_slides_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="atacseq_slides_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="atacseq_slides_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="atacseq_slides_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="atacseq_slides_files/libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="atacseq_slides_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="atacseq_slides_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="atacseq_slides_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="atacseq_slides_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="atacseq_slides_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': true,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'jumpToSlide': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleScrollView(event)\"><kbd>r</kbd> Scroll View Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"8\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1200,

        height: 800,

        // Factor of the display size that should remain empty around the content
        margin: 5.0e-2,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
      window.document.addEventListener("DOMContentLoaded", function (event) {
        const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
        tabsets.forEach(function(tabset) {
          const tabby = new Tabby('#' + tabset.id);
        });
        const isCodeAnnotation = (el) => {
          for (const clz of el.classList) {
            if (clz.startsWith('code-annotation-')) {                     
              return true;
            }
          }
          return false;
        }
        const onCopySuccess = function(e) {
          // button target
          const button = e.trigger;
          // don't keep focus
          button.blur();
          // flash "checked"
          button.classList.add('code-copy-button-checked');
          var currentTitle = button.getAttribute("title");
          button.setAttribute("title", "Copied!");
          let tooltip;
          if (window.bootstrap) {
            button.setAttribute("data-bs-toggle", "tooltip");
            button.setAttribute("data-bs-placement", "left");
            button.setAttribute("data-bs-title", "Copied!");
            tooltip = new bootstrap.Tooltip(button, 
              { trigger: "manual", 
                customClass: "code-copy-button-tooltip",
                offset: [0, -8]});
            tooltip.show();    
          }
          setTimeout(function() {
            if (tooltip) {
              tooltip.hide();
              button.removeAttribute("data-bs-title");
              button.removeAttribute("data-bs-toggle");
              button.removeAttribute("data-bs-placement");
            }
            button.setAttribute("title", currentTitle);
            button.classList.remove('code-copy-button-checked');
          }, 1000);
          // clear code selection
          e.clearSelection();
        }
        const getTextToCopy = function(trigger) {
            const codeEl = trigger.previousElementSibling.cloneNode(true);
            for (const childEl of codeEl.children) {
              if (isCodeAnnotation(childEl)) {
                childEl.remove();
              }
            }
            return codeEl.innerText;
        }
        const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
          text: getTextToCopy
        });
        clipboard.on('success', onCopySuccess);
        if (window.document.getElementById('quarto-embedded-source-code-modal')) {
          const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
            text: getTextToCopy,
            container: window.document.getElementById('quarto-embedded-source-code-modal')
          });
          clipboardModal.on('success', onCopySuccess);
        }
          var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
          var mailtoRegex = new RegExp(/^mailto:/);
            var filterRegex = new RegExp('/' + window.location.host + '/');
          var isInternal = (href) => {
              return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
          }
          // Inspect non-navigation links and adorn them if external
         var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
          for (var i=0; i<links.length; i++) {
            const link = links[i];
            if (!isInternal(link.href)) {
              // undo the damage that might have been done by quarto-nav.js in the case of
              // links that we want to consider external
              if (link.dataset.originalHref !== undefined) {
                link.href = link.dataset.originalHref;
              }
            }
          }
        function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
          const config = {
            allowHTML: true,
            maxWidth: 500,
            delay: 100,
            arrow: false,
            appendTo: function(el) {
                return el.closest('section.slide') || el.parentElement;
            },
            interactive: true,
            interactiveBorder: 10,
            theme: 'light-border',
            placement: 'bottom-start',
          };
          if (contentFn) {
            config.content = contentFn;
          }
          if (onTriggerFn) {
            config.onTrigger = onTriggerFn;
          }
          if (onUntriggerFn) {
            config.onUntrigger = onUntriggerFn;
          }
            config['offset'] = [0,0];
            config['maxWidth'] = 700;
          window.tippy(el, config); 
        }
        const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
        for (var i=0; i<noterefs.length; i++) {
          const ref = noterefs[i];
          tippyHover(ref, function() {
            // use id or data attribute instead here
            let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
            try { href = new URL(href).hash; } catch {}
            const id = href.replace(/^#\/?/, "");
            const note = window.document.getElementById(id);
            if (note) {
              return note.innerHTML;
            } else {
              return "";
            }
          });
        }
        const findCites = (el) => {
          const parentEl = el.parentElement;
          if (parentEl) {
            const cites = parentEl.dataset.cites;
            if (cites) {
              return {
                el,
                cites: cites.split(' ')
              };
            } else {
              return findCites(el.parentElement)
            }
          } else {
            return undefined;
          }
        };
        var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
        for (var i=0; i<bibliorefs.length; i++) {
          const ref = bibliorefs[i];
          const citeInfo = findCites(ref);
          if (citeInfo) {
            tippyHover(citeInfo.el, function() {
              var popup = window.document.createElement('div');
              citeInfo.cites.forEach(function(cite) {
                var citeDiv = window.document.createElement('div');
                citeDiv.classList.add('hanging-indent');
                citeDiv.classList.add('csl-entry');
                var biblioDiv = window.document.getElementById('ref-' + cite);
                if (biblioDiv) {
                  citeDiv.innerHTML = biblioDiv.innerHTML;
                }
                popup.appendChild(citeDiv);
              });
              return popup.innerHTML;
            });
          }
        }
      });
      </script>
    

</body></html>